### PR TITLE
Support EventChannel C++ plugin API for Linux/Windows

### PR DESCRIPTION
**Pull request by Kurun-pan**
#49398
https://github.com/flutter/flutter/issues/49398

* Tested on Linux desktop environment
* CI unit test wrote code, but it has not been confirmed

### CODE REVIEW

**Comment by Kurun-pan** [comment]
@googlebot I signed it!

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// Implementations must be prepared to accept sequences of alternating calls to
+// onListen(T, EventSink) and onCancel(Object). Implementations should ideally
+// consume no resources when the last such call is not onListen(). In typical
+// situations, this means that the implementation should register itself with
+// platform-specific event sources onListen() and deregister again onCancel().
+template <typename T>
+class StreamHandler {
+ public:
+  // Handles a request to set up an event stream.
+  // @param arguments stream configuration arguments, possibly null.
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h]: Why is this notation being used only for a subset of this one class?

I'm fine with adopting Doxygen for the client wrapper, but if it's going to be done it should be done consistently in all the new code.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  // @param event_sink an EventSink for emitting events to the Flutter receiver.
+  using OnListen =
+      std::function<void(const T* arguments, EventSink<T>* event_sink)>;
+
+  // Handles a request to tear down the most recently created event stream.
+  // @param arguments stream configuration arguments, possibly null.
+  using OnCancel = std::function<void(const T* arguments)>;
+
+  OnListen onListen;
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h]: ivars may not be public per C++ style guide.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  OnCancel onCancel;
+
+  // Registers a stream handler on this channel.
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h]: This comment is copypasta.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+          std::copy(result->begin(), result->end(), buffer);
+          reply(buffer, result->size());
+          delete[] buffer;
+        }
+      } else if (method.compare(kOnCancelMethod) == 0) {
+        if (current_sink) {
+          handler.onCancel(method_call->arguments());
+
+          auto result = codec->EncodeSuccessEnvelope();
+          uint8_t* buffer = new uint8_t[result->size()];
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h]: Note that you are leaking `buffer` here, demonstrating why using `new`/`delete` manually should be avoided whenever possible. (Separately from the fact that I don't think this buffer should exist in the first place.)

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_sink.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_COMMON_CPP_CLIENT_WRAPPER_INCLUDE_FLUTTER_EVENT_SINK_H_
+#define FLUTTER_SHELL_PLATFORM_COMMON_CPP_CLIENT_WRAPPER_INCLUDE_FLUTTER_EVENT_SINK_H_
+
+namespace flutter {
+
+// Event callback. Supports dual use: Producers of events to be sent to Flutter
```
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_sink.h]: Modify comment

```suggestion
// Event callback. Events to be sent to Flutter application
// act as clients of this interface for sending events.
```

**Comment by Kurun-pan** [comment]
@stuartmorgan 
Thank you for your review. I'll confirm and fix.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 1 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h]: I used the comments of following source code for reference. I thought the same comment would be better.
`shell/platform/android/io/flutter/plugin/common/EventChannel.java`

There's definitely no sense of unity in the code. Adjust to other C++ code, I try not to use Doxygen Doc comments.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 2 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h]: Yes, you're right. But, I have no idea of preparing the API compatible with [setStreamHandler(Java API)](https://api.flutter.dev/javadoc/io/flutter/plugin/common/EventChannel.html#setStreamHandler-io.flutter.plugin.common.EventChannel.StreamHandler-).

Do you have any ideas?

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 3 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h]:
```suggestion
```

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h (hunk 4 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h]: These code will be removed in the first place.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        }
+
+        current_sink =
+            new EventSinkImplementation(messenger, channel_name, codec);
+        handler.onListen(method_call->arguments(),
+                         static_cast<EventSink<T>*>(current_sink));
+
+        {
+          auto result = codec->EncodeSuccessEnvelope();
+          uint8_t* buffer = new uint8_t[result->size()];
```
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h]: No copy was needed. Delete it.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+  EventChannel& operator=(EventChannel const&) = delete;
+
+  // Registers a stream handler on this channel.
+  // If no handler has been registered, any incoming stream setup requests will
+  // be handled silently by providing an empty stream.
+  void SetStreamHandler(const StreamHandler<T>& handler) const {
+    // TODO: The following is required when nullptr
+    // can be passed as an argument.
+    // if (!handler) { /* <= available for more than C++17 */
+    //   messenger_->SetMessageHandler(name_, nullptr);
```
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h]: There is no way. I think again.

**Comment by stuartmorgan** [comment]
> Is it okay to commit multiple times?

To the PR? Sure, feel free to do incremental commits as is easiest for you. I'd just ask that you not force-push any rebases if you can avoid it, since GitHub's review system doesn't handle that very well.

**Comment by Kurun-pan** [comment]
@stuartmorgan 

Thank you for your reply.

>I'd just ask that you not force-push any rebases if you can avoid it, since GitHub's review system doesn't handle that very well.

Sure. 
I'll fix these codes as soon as possible.

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+  // Send a test message to trigger the handler test assertions.
+  MethodCall<EncodableValue> call("listen", nullptr);
+  auto message = codec.EncodeMethodCall(call);
+  messenger.last_message_handler()(
+      message->data(), message->size(),
+      [](const uint8_t* reply, const size_t reply_size) {});
+  EXPECT_EQ(on_listen_called, true);
+
+  // FIXME: add onCancel test scenario
```
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc]: I fixed unit test code.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h]: I implemented it.

**Comment by Kurun-pan** [comment]
@stuartmorgan

I'm sorry for too late. Could you review again?

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+// Tests that SetStreamHandler sets a handler that correctly interacts with
+// the binary messenger.
+TEST(EventChannelTest, Registration) {
+  TestBinaryMessenger messenger;
+  const std::string channel_name("some_channel");
+  const StandardMethodCodec& codec = StandardMethodCodec::GetInstance();
+  EventChannel channel(&messenger, channel_name, &codec);
+
+  bool on_listen_called = false;
+  auto onListen =
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:57]: `on_listen`

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      [&on_listen_called](
+          const flutter::EncodableValue* arguments,
+          std::unique_ptr<flutter::EventSink<flutter::EncodableValue>>&&
+              events) { on_listen_called = true; };
+  auto onCancel = [](const flutter::EncodableValue* arguments) {};
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:62]: `on_cancel`

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+  flutter::StreamHandler<flutter::EncodableValue> handler(onListen, onCancel);
+  channel.SetStreamHandler(&handler);
+  EXPECT_EQ(messenger.last_message_handler_channel(), channel_name);
+  EXPECT_NE(messenger.last_message_handler(), nullptr);
+
+  // Send dummy listen message
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:69]: Nit: Missing a period at the end of the sentence (per style guide)

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+  MethodCall<flutter::EncodableValue> call("listen", nullptr);
+  auto message = codec.EncodeMethodCall(call);
+  messenger.last_message_handler()(
+      message->data(), message->size(),
+      [](const uint8_t* reply, const size_t reply_size) {});
+
+  // Check results
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:76]: Same

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc (hunk 10 above)
Comment (stuartmorgan) [reply shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:62]: These both need to be fixed in the other test methods as well.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+// A named channel for communicating with the Flutter application using
+// asynchronous event streams. Incoming requests for event stream setup are
+// decoded from binary on receipt, and C++ responses and events are encoded into
+// binary before being transmitted back to Flutter. The MethodCodec used must be
+// compatible with the one used by the Flutter application. This can be achieved
+// by creating an EventChannel
+// ("https://docs.flutter.io/flutter/services/EventChannel-class.html")
+// counterpart of this channel on the Dart side.
+// The C++ type of stream configuration arguments, events, and error details are
+// Template, but only values supported by the specified MethodCodec can be used.
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:28]: s/Template/templated/

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+        auto result = codec->EncodeSuccessEnvelope();
+        reply(result->data(), result->size());
+      } else if (method.compare(kOnCancelMethod) == 0) {
+        handler->onCancel(method_call->arguments());
+
+        auto result = codec->EncodeSuccessEnvelope();
+        reply(result->data(), result->size());
+      } else {
+        std::cerr
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:86]: Logging unhandled methods should be handled on the Flutter side already, so this shouldn't be needed.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+                  << channel_name << std::endl;
+        return;
+      }
+
+      const std::string& method = method_call->method_name();
+      if (method.compare(kOnListenMethod) == 0) {
+        auto sink = std::make_unique<EventSinkImplementation>(
+            messenger, channel_name, codec);
+        handler->onListen(method_call->arguments(), std::move(sink));
+
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:77]: Why isn't there any concept/handling of a current sink here? This class doesn't seem to behave at all like the ObjC or Java implementations.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_sink.h [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+  EventSink() = default;
+  virtual ~EventSink() = default;
+
+  // Prevent copying.
+  EventSink(EventSink const&) = delete;
+  EventSink& operator=(EventSink const&) = delete;
+
+  // Consumes end of stream. Ensuing calls to Success() or
+  // Error(), if any, are ignored.
+  void EndOfStream() { EndOfStreamInternal(); }
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_sink.h:24]: Nit: it seems like it would be clear to make this the last method, rather than the first.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_sink.h [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  // Consumes an error event.
+  void Error(const std::string& error_code,
+             const std::string& error_message = "",
+             T* error_details = nullptr) {
+    ErrorInternal(error_code, error_message, error_details);
+  }
+
+ protected:
+  // Implementation of the public interface, to be provided by subclasses.
+  virtual void EndOfStreamInternal() = 0;
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_sink.h:38]: Same.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+namespace flutter {
+
+// Handler of stream setup and tear-down requests.
+// Implementations must be prepared to accept sequences of alternating calls to
+// onListen() and onCancel(). Implementations should ideally consume no
+// resources when the last such call is not onListen(). In typical situations,
+// this means that the implementation should register itself with
+// platform-specific event sources onListen() and deregister again onCancel().
+template <typename T>
+struct StreamHandler {
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:19]: What is the argument for this being a struct of function objects rather than—as on the other platforms—an interface? Using different designs for the different language variants adds some conceptual overhead for people dealing with multiple variants, so I'd like to be sure there's a compelling reason for this difference.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h (hunk 15 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:84]: @stuartmorgan 
I fixed it to be the same behaviour like the Java and ObjC implementations.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 18 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:19]: @stuartmorgan 

At the point of [the review](https://github.com/flutter/engine/pull/17015#discussion_r391850887) , I changed from class to struct. 

In addition, MethodChannel is implemented using function, so I thought that it would be better to use function in the same way. I could only think of this to use function with one argument. Is it better to use Interface Class? 

I hope this answers your question.

[method_channel.h](https://github.com/flutter/engine/blob/master/shell/platform/common/cpp/client_wrapper/include/flutter/method_channel.h)
```
using MethodCallHandler =
    std::function<void(const MethodCall<T>& call,
                       std::unique_ptr<MethodResult<T>> result)>;

void SetMethodCallHandler(MethodCallHandler<T> handler) const {
```

**Comment by Kurun-pan** [comment]
@stuartmorgan

Thank you for your review.
I have finished the fix. Could you check the source code and my reply?

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 18 above)
Comment (stuartmorgan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:19]: Since the last review, I needed to address the same thing for MethodResult. For consistency, please follow the pattern I used there of having [an interface class](https://github.com/flutter/engine/blob/9992a4ae664b0de4bf1834ad44dc48c7ee6025e8/shell/platform/common/cpp/client_wrapper/include/flutter/method_result.h), and then a [function-based implementation](https://github.com/flutter/engine/blob/9992a4ae664b0de4bf1834ad44dc48c7ee6025e8/shell/platform/common/cpp/client_wrapper/include/flutter/method_result_functions.h) for cases where implementing the interface isn't convenient.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+    if (!handler) {
+      messenger_->SetMessageHandler(name_, nullptr);
+      is_listened_ = false;
+      return;
+    }
+
+    const MethodCodec<T>* codec = codec_;
+    const std::string channel_name = name_;
+    const BinaryMessenger* messenger = messenger_;
+    bool& is_listend = is_listened_;
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:58]: I'm not seeing what purpose this layer of references is serving. If you want to modify the ivar in the lamba, you should capture `this`.

Except looking at the rest of the code, and at the ObjC implementation, I'm not sure why this is an ivar at all. The state is only needed in association with a given handler, as far as I can tell, in which case you can make this a local variable captured by the lambda, and eliminate the ivar entirely (which is how the ObjC version is structured).

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+    BinaryMessageHandler binary_handler = [handler, codec, channel_name,
+                                           messenger, &is_listend](
+                                              const uint8_t* message,
+                                              const size_t message_size,
+                                              BinaryReply reply) {
+      constexpr char kOnListenMethod[] = "listen";
+      constexpr char kOnCancelMethod[] = "cancel";
+
+      std::shared_ptr<MethodCall<T>> method_call =
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:67]: I don't see any reason for this to be a `shared_ptr` rather than a `unique_ptr`.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+          codec->DecodeMethodCall(message, message_size);
+      if (!method_call) {
+        std::cerr << "Unable to construct method call from message on channel: "
+                  << channel_name << std::endl;
+        return;
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:72]: You aren't calling `reply` on this codepath, which violates the API contract.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h (hunk 19 above)
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:58]: Nit: this should be called `is_listening`

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+  // Handles a request to set up an event stream.
+  // |arguments| is stream configuration arguments and
+  // |events| is an EventSink for emitting events to the Flutter receiver.
+  using OnListen = std::function<void(const T* arguments,
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:23]: These need a way to communicate errors, as with the Java and ObjC versions—again, please don't deviate from the existing versions of these APIs with a strong justification.

Since we don't already have an error object, I think you'll need to make a simple `StreamHandlerError` object with the same arguments as an error envelope needs, and have these return a (possibly null) `std::unique_ptr<StreamHandlerError<T>>`. (It would be nice if we could have a common error object like ObjC, but this version will need the details to be a `std::unique_ptr<T>`, and the error calls take `const T*` for cases where it's not efficient to pass ownership, so that's not workable.)

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+
+  // Send dummy listen message.
+  MethodCall<flutter::EncodableValue> call("listen", nullptr);
+  auto message = codec.EncodeMethodCall(call);
+  messenger.last_message_handler()(
+      message->data(), message->size(),
+      [](const uint8_t* reply, const size_t reply_size) {});
+  EXPECT_EQ(on_listen_called, true);
+
+  // Send dummy listen message.
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:168]: Please expand on this comment to explain why you are sending it again.

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+  message = codec.EncodeMethodCall(call);
+  messenger.last_message_handler()(
+      message->data(), message->size(),
+      [](const uint8_t* reply, const size_t reply_size) {});
+  EXPECT_EQ(on_listen_called, true);
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:173]: You didn't set it to false, after the previous assert, so this line isn't actually testing anything.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h (hunk 20 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:67]: Sorry, this is my mistake.  `unique_ptr` is correct.

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc (hunk 23 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:168]: I modified the comment.

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc (hunk 24 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:171]: Thank you. I added `on_listen_called` reset.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 18 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:19]: OK, thank you.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h (hunk 21 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:72]: I added reply.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h (hunk 19 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:58]: I modified from `is_listend` to `is_listening`.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 22 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:23]: > These need a way to communicate errors, as with the Java and ObjC versions—again, please don't deviate from the existing versions of these APIs with a strong justification.

At first, I thought that Java (EventCHannel.java) and Objective-C (FlutterChannels.mm) had different specifications. But when I thought about it, I understood what you said. Thank you.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h (hunk 19 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h]: I modified, but I'm still using `is_listened_`. 
I think it's simpler to use the flag (`is_listened_`), but could you give any advice if there is another good implementation?

**Comment by Kurun-pan** [comment]
@stuartmorgan 

Thank you for your review.

> Some comments still, but I think this is very close. Thanks for your work on this!
I'm glad to hear that! Thank you.

I'm done fixing. Could you request again?

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+ public:
+  StreamHandler() = default;
+  virtual ~StreamHandler() = default;
+
+  // Prevent copying.
+  StreamHandler(StreamHandler const&) = delete;
+  StreamHandler& operator=(StreamHandler const&) = delete;
+
+  // Handles a request to set up an event stream. Returns error if representing
+  // an unsuccessful outcome of invoking the method, possibly nullptr.
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:43]: I'm finding this sentence a bit hard to parse. How about just "Returns nullptr on success, or an error on failure."

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+  // |arguments| is stream configuration arguments and
+  // |events| is an EventSink for emitting events to the Flutter receiver.
+  std::unique_ptr<StreamHandlerError<T>> OnListen(
+      const T* arguments,
+      std::unique_ptr<EventSink<T>>&& events) {
+    return OnListenInternal(arguments, std::move(events));
+  }
+
+  // Handles a request to tear down the most recently created event stream.
+  // Returns error if representing an unsuccessful outcome of invoking the
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:53]: Same here.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler_functions.h [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+  // |flutter::StreamHandler|
+  std::unique_ptr<StreamHandlerError<T>> OnListenInternal(
+      const T* arguments,
+      std::unique_ptr<EventSink<T>>&& events) override {
+    if (on_listen_) {
+      return on_listen_(arguments, std::move(events));
+    }
+
+    auto error = std::make_unique<StreamHandlerError<T>>(
+        "error", "Not found StreamHandlerListen hander", nullptr);
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler_functions.h:54]: "No OnListen handler set"

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler_functions.h [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+
+  // |flutter::StreamHandler|
+  std::unique_ptr<StreamHandlerError<T>> OnCancelInternal(
+      const T* arguments) override {
+    if (on_cancel_) {
+      return on_cancel_(arguments);
+    }
+
+    auto error = std::make_unique<StreamHandlerError<T>>(
+        "error", "Not found StreamHandlerCancel hander", nullptr);
```
Comment (stuartmorgan) [inline shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler_functions.h:66]: "No OnCancel handler set"

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler_functions.h (hunk 27 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler_functions.h:54]: Modified.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler_functions.h (hunk 28 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler_functions.h:66]: Modified.

File: shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+  message = codec.EncodeMethodCall(call_cancel);
+  messenger.last_message_handler()(
+      message->data(), message->size(),
+      [](const uint8_t* reply, const size_t reply_size) {});
+
+  // Check results.
+  EXPECT_EQ(on_cancel_called, true);
+}
+
+// Test that consecutive call of OnListen.
```
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/event_channel_unittests.cc:155]: Modified.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 30]
```diff
@@ -0,0 +1,169 @@
...
+        std::cerr << "Unable to construct method call from message on channel: "
+                  << channel_name << std::endl;
+        reply(nullptr, 0);
+        return;
+      }
+
+      const std::string& method = method_call->method_name();
+      if (method.compare(kOnListenMethod) == 0) {
+        if (is_listening_) {
+          auto error = shared_handler->OnCancel(nullptr);
```
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:82]: Modified.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h [hunk 31]
```diff
@@ -0,0 +1,169 @@
...
+                      << (error->error_code) << ", " << (error->error_message)
+                      << ", " << (error->error_details);
+          }
+        }
+        is_listening_ = true;
+
+        std::unique_ptr<std::vector<uint8_t>> result;
+        auto sink = std::make_unique<EventSinkImplementation>(
+            messenger, channel_name, codec);
+        auto error =
```
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h:94]: Modified.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+                     const std::string& error_message,
+                     const T* error_details)
+      : error_code(error_code),
+        error_message(error_message),
+        error_details(error_details) {}
+};
+
+// Handler of stream setup and tear-down requests.
+// Implementations must be prepared to accept sequences of alternating calls to
+// onListen() and onCancel(). Implementations should ideally consume no
```
[the remark this answers is not in the document]
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:28]: Modified.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 25 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:43]: Thank you! Modified.

File: shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h (hunk 26 above)
Comment (Kurun-pan) [reply shell/platform/common/cpp/client_wrapper/include/flutter/event_stream_handler.h:53]: Thank you! Modified.

**Comment by Kurun-pan** [comment]
@stuartmorgan 
Thank you for your review. I fixed. Could you check again? Thank you as always.

**Comment by Kurun-pan** [comment]
Thank you so much!

**Comment by beevelop** [comment]
@Kurun-pan Thank you man for the PR. Awesome stuff 🚀

**Comment by Kurun-pan** [comment]
@beevelop 
Thank you for your message. I'm very happy to get your message.

**Comment by matt-deboer** [comment]
This looks exciting; are there any examples out there that leverage this API yet for a Linux plugin?

**Comment by stuartmorgan** [comment]
Linux switched to a completely different API surface several months ago, so the C++ API is now Windows-only. I've filed https://github.com/flutter/flutter/issues/65270 for adding it to the new Linux API.

**Comment by DlmuZQ** [comment]
How do I use the event channel on windows? Can you give some example?

**Comment by fawdlstty** [comment]
I have try to use EventChannel, but I found memory leaks and crashes. Question and example code here: [#76448](https://github.com/flutter/flutter/issues/76448)

who can give some help? thanks in advance
